Make use of pytest for all tests

okin opened the pull request:
### Pull Request Checklist

- [x] I’ve read the [guidelines for contributing](https://github.com/getnikola/nikola/blob/master/CONTRIBUTING.rst).
- [x] I updated AUTHORS.txt and CHANGES.txt (if the change is non-trivial) and documentation (if applicable).
- [x] I tested my changes.

### Description

As part of Hacktoberfest I started to look into the tests of Nikola and found that they...

* use a mixture of test styles (xUnit-style tests with unittest.TestCase; pytest functions)
* test more things than one thing in one test
* not all tests have been running
* had side-effects
* sometimes relied on other tests to be run first

I decided to port all the tests to pytest.
Pytest has been used as a test runner for some while now and provides more features that come in handy.
Pytest tests are written as simple functions which makes writing a test easier.
In combination with fixtures we are able to reduce the amount of setup code and we are able to effictively reduce side effects by using fixtures with [autouse](https://docs.pytest.org/en/latest/fixture.html#autouse-fixtures-xunit-setup-on-steroids).

This PR makes use of test parametrization and fixtures.
A lot of the tests have been refactored beyond a simple switch in test style to make them more explicit and easier to understand.

Kwpolska commented on tests/test_plugin_importing.py:
```diff
@@ -1,14 +1,2 @@
-
-    def test_importing_plugin_compile_markdown(self):
-        import nikola.plugins.compile.markdown    # NOQA
+def test_importing_plugin_task_galleries():
```
This could be merged with the equally short `test_commands.py`.

okin replied:
Merged them into `test_plugins.py`.

Kwpolska commented on tests/test_rst_compiler.py:
```diff
@@ -1,238 +1,227 @@
+
+
+def test_ReST_extension(tempdir):
+    "Simple test for our base class :)"
```
The docstring is probably outdated.  Also, we prefer `"""triple quotes"""` for docstrings.

okin replied:
There will be a few other places where single quotes are in place. I'll check for these!

Kwpolska replied:
You could try to use the `black` automatic formatter on tests only. It’s a great tool overall; converting all of Nikola to `black` would make a huge mess, but since you’re editing practically all files under `tests/`, it will work out here nicely. Feel free to pick a length limit greater than the default (88).

Kwpolska commented on tests/test_rst_compiler.py:
```diff
@@ -1,238 +1,227 @@
-                                attributes={"src": ("https://player.vimeo.com/"
-                                                    "video/VID"),
-                                            "height": "400", "width": "600"})
+def getHtmlFromRst(temp_dir, rst):
```
Please use `snake_case` here and in `assertHTMLContains`.

okin commented:
Work on this is still ongoing but nearing completion.

In the meantime I have split up the larger integration test module into separate modules. During this process I also added the tests that run the check commands to most integration tests as I couldn't think of why any of these tested settings should not be able to get a positive check result.
This did increase the runtime of the tests and if this is of concern the calls to the check commands could be removed from the tests.

There was one test for doing a wordpress import which I have converted to an automated tests aswell. This resides alongside the other integration tests.

I was also able to remove some appearantly old or unused parts of the test setups and now hopefully everything that is in the tests is actually required to run them.

My next step is working on formatting - probably with help of `black`.

Kwpolska commented on tests/test_rst_compiler.py:
```diff
@@ -1,238 +1,237 @@
+    compiler = nikola.plugins.compile.rest.CompileRest()
+    compiler.set_site(FakeSite())
+    compiler.site.post_per_input_file[infile] = FakePost(outfile)
+    compiler.compile(infile, outfile)
```
Couldn’t we just call `compiler.compile_string` and avoid dealing with temp files?

okin replied:
We get one rather unimportant line less covered but have less setup required for tests.
I'd say this is a good tradeoff!

okin commented on tests/test_command_import_wordpress.py:
```diff
@@ -1,294 +1,116 @@
+    write_attachments_info = mock.MagicMock()
+    download_mock = mock.MagicMock()
+
+    with mock.patch(
```
It works totally fine without.
Removed the nesting.

Kwpolska commented on tests/integration/test_check_failure.py:
```diff
@@ -0,0 +1,51 @@
+"""The demo build should pass 'nikola check' and fail with missing files."""
```
Where is it tested that the demo site passes `nikola check`?

okin replied:
This can be found in `integration/test_demo_build.py`.

Kwpolska replied:
It isn’t obvious without reading/being aware of `test_demo_build.py`.

okin replied:
I get your point and added some documentation. Let me know what you think about this!

Kwpolska replied:
That’s more obvious, thanks.

Kwpolska commented on tests/integration/test_redirection_absolute.py:
```diff
@@ -0,0 +1,49 @@
+)
+
+
+@pytest.fixture(scope="module")
```
Please restructure this.

1. Merge all three `test_redirection_*.py` files into one (create all three redirects in one site)
2. Write a test to check that each of the three redirects works (file is created and has a meta refresh tag with the appropriate URL)

okin replied:
I've merged all redirection tests into one file (`tests/integration/test_redirections.py`) and added the tests.
What do you say about it now?

Kwpolska replied:
Much better.

Kwpolska commented on tests/integration/test_section_page_collision.py:
```diff
@@ -0,0 +1,96 @@
+    Add subdirectories and create a post in section "sec1" and a page
+    with the same URL as the section index.
+
+    It also enables post sections.
```
[Sections were a feature that was dropped in Nikola v8.](https://getnikola.com/blog/upgrading-to-nikola-v8.html#sections-were-replaced-by-categories) This test hasn’t been doing anything interesting for quite a while. That said, the layout created by sections can be recreated by `CATEGORY_DESTPATH_AS_DEFAULT = True; CATEGORY_PAGES_FOLLOW_DESTPATH = True`. It might be useful to activate that with `patch_config` in this test and rename it.

okin replied:
Nikola failed me with the various attempts I had a re-enabling the configuration and make the test do something useful.

Do you have any objections in dropping this test module alltogether?

Kwpolska replied:
```
$ tree posts
posts
├── 1.rst
├── a
│   ├── 1.rst
│   └── 2.rst
├── b
│   └── 1.rst
└── b.rst
```

conf.py changes

```python
CATEGORY_ALLOW_HIERARCHIES = True
CATEGORY_DESTPATH_AS_DEFAULT = True
CATEGORY_DESTPATH_TRIM_PREFIX = True
CATEGORY_PAGES_FOLLOW_DESTPATH = True
```

results:
* `/output/posts/a/index.html` is a category index (“Posts about a”) with two posts
* `/output/posts/b/index.html`… crashes with current master, so that needs fixing

I’ll take it from there, fixing both the code and the test, or dropping it if it’s impossible.

okin commented on tests/integration/test_building_in_subdir.py:
```diff
@@ -0,0 +1,45 @@
+
+
+@pytest.fixture(scope="module")
+def build_dir(target_dir):
```
Nope, not anymore. If it will be needed in the future it can easily be re-introduced.

okin commented on tests/integration/test_archive_full.py:
```diff
@@ -0,0 +1,53 @@
+
+@pytest.fixture(scope="module")
+def build(target_dir, test_dir):
+    """Fill the site with demo content and build it."""
```
I moved the setup of the demo site into a function that is re-used through the tests that use it. What do you think?

Kwpolska replied:
It’s much better now!

Kwpolska commented on tests/integration/test_building_in_subdir.py:
```diff
@@ -0,0 +1,33 @@
+
+    with cd(build_dir):
+        result = __main__.main(["build"])
+        assert result == 0
```
This `assert` does not appear anywhere else, why is it there?

okin replied:
It's been like this for ages and I think it has been used to check for a valid build since the build is triggered in a different directory.

Since we are employing more tests now that check for valid builds (and fail if no proper build has been made) I'd say we remove it.
What do you think?

Kwpolska replied:
If a non-zero exit code will guarantee other tests in this file failing, and if that’s the case for all tests where we build something, then this `assert` can be dropped.

Kwpolska commented:
LGTM! Before merging, please resolve `tests/integration/test_building_in_subdir.py` (remove the assert if that’s safe). And if you wish to do any sort of rebasing/commit squashing/history simplification, now is the time. (It’s not necessary.)
